lib/api/CBoostedTreeInferenceModelBuilder.cc

```diff
@@ -0,0 +1,193 @@
+    if (inputColumnIndex < m_Definition.input().fieldNames().size()) {
+        std::string fieldName{m_Definition.input().fieldNames()[inputColumnIndex]};
+        std::string featureName{fieldName + "_frequency"};
+        auto stringMap = this->encodingMap(inputColumnIndex, map);
+        m_Definition.preprocessors().push_back(std::make_unique<CFrequencyEncoding>(
+            fieldName, featureName, std::move(stringMap)));
+        m_FeatureNames.push_back(featureName);
+    }
+}
+
+CInferenceModelDefinition&& CBoostedTreeInferenceModelBuilder::build() {
+
+    // Finalize OneHotEncoding Mappings
+    for (auto& oneHotEncodingMapping : m_OneHotEncodingMaps) {
+        m_Definition.preprocessors().emplace_back(
+            std::move(oneHotEncodingMapping.second));
+    }
+
+    // Add aggregated output after the number of trees is known
+    auto ensemble{static_cast<CEnsemble*>(m_Definition.trainedModel().get())};
+    this->setAggregateOutput(ensemble);
+
+    this->setTargetType();
+    ensemble->featureNames(m_FeatureNames);
+
+    return std::move(m_Definition);
+}
+
+void CBoostedTreeInferenceModelBuilder::addNode(
+    std::size_t splitFeature,
+    double splitValue,
+    bool assignMissingToLeft,
+    double nodeValue,
+    double gain,
+    ml::maths::CBoostedTreeNode::TOptionalNodeIndex leftChild,
+    ml::maths::CBoostedTreeNode::TOptionalNodeIndex rightChild) {
+    auto ensemble{static_cast<CEnsemble*>(m_Definition.trainedModel().get())};
+    // use dynamic cast to prevent using wrong type of trained models
+    auto tree = dynamic_cast<CTree*>(ensemble->trainedModels().back().get());
+    if (tree == nullptr) {
+        HANDLE_FATAL(<< "Internal error. Tree points to a nullptr.")
+    }
+    tree->treeStructure().emplace_back(tree->size(), splitValue, assignMissingToLeft, nodeValue,
+                                       splitFeature, leftChild, rightChild, gain);
+}
+
+CBoostedTreeInferenceModelBuilder::CBoostedTreeInferenceModelBuilder(
+    TStringVec fieldNames,
+    std::size_t dependentVariableColumnIndex,
+    const TStringSizeUMapVec& categoryNameMap) {
+    // filter filed names containing only "."
+    fieldNames.erase(std::remove(fieldNames.begin(), fieldNames.end(), "."),
+                     fieldNames.end());
+    // filter dependent variable field name
+    assert(dependentVariableColumnIndex < fieldNames.size());
+    fieldNames.erase(fieldNames.begin() +
+                     static_cast<std::ptrdiff_t>(dependentVariableColumnIndex));
+    m_FieldNames = fieldNames;
+
+    this->categoryNameMap(categoryNameMap);
+    m_Definition.fieldNames(fieldNames);
+    m_Definition.trainedModel(std::make_unique<CEnsemble>());
+    m_Definition.typeString(INFERENCE_MODEL);
+}
+
+CBoostedTreeInferenceModelBuilder::TStringDoubleUMap
+CBoostedTreeInferenceModelBuilder::encodingMap(std::size_t inputColumnIndex,
+                                               const TDoubleVec& map_) {
+    TStringDoubleUMap map;
```

Comment:
Question: do we ever actually need to lookup in this map or is it just for outputting as name value pairs in the JSON? If it is the latter, maybe we should switch the type `std::vector<std::pair<std::string, double>>`: it would be more space and memory efficient.

Reply:
I changed the type to `TStrVecVec`.